dune/iga/nurbstrimfunctionality.hh

```diff
@@ -0,0 +1,624 @@
+
+  bool approxSamePoint(ClipperPoint a, ClipperPoint b, const double tol = tolerance) {
+    return Dune::FloatCmp::eq(distance(a, b), 0.0, tol);
+  }
+
+  int getEdgeOrientation(const int edge) { return (edge == 0 || edge == 2) ? 0 : 1; }
+
+  int iPlusX(int i, int x) { return (i + x) % 4; }
+
+  std::array<Point, 2> curveStartEndPoint(const CurveGeometry& curve) {
+    std::array<int, 2> indices{0, static_cast<int>(curve.patchData_->controlPoints.size()[0]) - 1};
+    return {curve.patchData_->controlPoints.get({indices[0]}).p, curve.patchData_->controlPoints.get({indices[1]}).p};
+  }
+
+  PointVector getElementCorners(const auto& element) {
+    auto geo = element.geometry();
+
+    std::vector<Point> corners(4);
+    corners[0] = geo.corner(0);
+    corners[1] = geo.corner(1);
+    corners[2] = geo.corner(3);  // see dune book page 127 Figure 5.12
+    corners[3] = geo.corner(2);
+
+    return corners;
+  }
+
+  Clipper2Lib::PathD getElementEdges(const auto& element) {
+    auto corners = getElementCorners(element);
+    Clipper2Lib::PathD edges;
+    for (int i = 0; i < 4; ++i)
+      edges.emplace_back(corners[i][0], corners[i][1]);
+    return edges;
+  }
+
+  Clipper2Lib::PathD getElementEdges(const PointVector& corners) {
+    Clipper2Lib::PathD edges;
+    for (int i = 0; i < 4; ++i)
+      edges.emplace_back(corners[i][0], corners[i][1]);
+    return edges;
+  }
+
+  Clipper2Lib::PathsD getClip(std::vector<Boundary>& boundaries) {
+    Clipper2Lib::PathD pathD;
+    for (auto& boundary : boundaries)
+      for (auto& point : boundary.path(pathSamples))
+        pathD.push_back(point);
+
+    // Trim Patch, so we don't get duplicate intersection Points on corners
+    pathD = Clipper2Lib::TrimCollinear(pathD, clipperPrecision);
+
+    return {pathD};
+  }
+
+  /// Checks if a point p is on the line spanned by the points a and b within a given tolerance
+  bool pointOnLine(const ClipperPoint& p, const ClipperPoint& a, const ClipperPoint& b, const double tol = tolerance) {
+    // https://stackoverflow.com/a/17693146
+    return Dune::FloatCmp::eq(distance(a, p) + distance(b, p), distance(a, b), tol);
+  }
+
+  std::pair<bool, int> pointOnAnyEdge(const ClipperPoint& point, const auto& elementEdges) {
+    for (int i = 0; i < 4; ++i)
+      if (pointOnLine(point, elementEdges[i], elementEdges[iPlusX(i, 1)])) return std::make_pair(true, i);
+
+    return std::make_pair(false, -1);
+  }
+
+  std::pair<bool, int> pointOnAnyVertex(const ClipperPoint& point, const PointVector& corners) {
+    for (size_t i = 0; i < corners.size(); ++i)
+      if (approxSamePoint(corners[i], point)) return std::make_pair(true, static_cast<int>(i));
+
+    return std::make_pair(false, -1);
+  }
+
+  // Sorts the intersectinPoints countr-clockwise arund the element (crucial step!)
+  void sortIntersectionPoints(const IntersectionPointMapPtr& pointMapPtr, const int edge) {
```

Comment:
as mentoined above, you now have here a `const` object, but you modify it by sort.
If the name would be `const std::shared_ptr<IntersectionPointMap>`, one would not wonder, since then it is what is const and what not.

Reply:
Genral question:
Is is better to pass a smart pointer as  reference or by value?

Reply:
if you dont take owner ship in the function in somw wa you should pass a raw reference or a raw pointer. otherwise if you take ownership you can pass by reference. but maybe we discuss this in person